### ISSUE TITLE
Bug: Envoy regenerates TraceID when the trace is not sampled, breaking observability flows

### ISSUE THREAD

**Issue by mkielar** [aws/aws-app-mesh-roadmap#411]
*SECURITY NOTICE: If you think you’ve found a potential security issue, please do not post it in the Issues. Instead, please follow the instructions [here](https://aws.amazon.com/security/vulnerability-reporting/) or [email AWS security directly](mailto:aws-security@amazon.com).*

**Summary**
We're running a bunch of ECS Fargate Services with AppMesh. We use Envoy as Service Sidecars, and as Virtual Gateway. We also use API GW as ingress. The infrastructure looks more or less like this:

API GW => VPC Link (NLB) => Virtual Gateway => ECS Services

We have a custom Sampling Rule in X-Ray, called "prod", which is configured with:
- Reservoir size: 1rps
- Fixed rate:     5%

The same is configured for all Envoy Sidecars in our AppMesh, using the `XRAY_SAMPLING_RULE_MANIFEST` environment variable. The manifest file looks like this for Virtual Gateway Envoy, and all Envoy Sidecars:
```
    {
        "default": {
            "fixed_target": 1,
            "rate": 0.05
        },
        "rules": [
            {
                "description": "Disable localhost tracing",
                "fixed_target": 0,
                "host": "localhost",
                "http_method": "*",
                "rate": 0,
                "url_path": "*"
            }
        ],
        "version": 2
    }
```
All our envoys are based on `public.ecr.aws/appmesh/aws-appmesh-envoy:v1.21.1.2-prod` image.

We have noticed that - when the API GW makes a decision of "not sampling" a trace - Envoys tend to regenerate TraceIDs, breaking observability. We believe this is a bug in X-Ray integration implementation of Envoy.

**Steps to Reproduce**
To prove/reproduce this, first I ran a single request against our API GW to get a "working" value for X-Amzn-Trace-Id:
```
curl -v https://api.example.com/my-service/v1/test-endpoint
...
< x-amzn-trace-id: Root=1-62833b4c-4a0cb5dd511bf2cd48f99580
```

Having that, I opened a terminal session to my Bastion Host, to be able to bypass API GW, and hit my endpoints directly. I ran all the tests against the same ECS Service, which is a .NET Application. The Service has logging configured to add X-Amzn-Trace-Id to all log entries, for distributed tracing. This usually enables us to correlate logs from different services even if the trace is not stored in X-Ray. Or, at least, that's what we thought...

I tested two routes:

### Route 1. Via Virtual Gateway.

The query:
```
curl -v -H "x-amzn-trace-id: <header value, see below>" vg.prod.local:8080/my-service/v1/test-endpoint
```

In this case, the nodes involved looked like this:

```
Client (Bastion) => Envoy (Virtual Gateway) => Envoy (service sidecar) => service
```

#### Sampled test:

Request:
```
curl -v -H "x-amzn-trace-id: Root=1-62833b4c-4a0cb5dd511bf2cd48f99580;Sampled=1" vg.prod.local:8080/my-service/v1/test-endpoint
```
Got in "my-service": 
```
Root=1-62833b4c-4a0cb5dd511bf2cd48f99580;Parent=ff32548f0b478d62;Sampled=1
```

Observation: 
When `Sampled=1`, Envoy uses the TraceID provided. This is correct behaviour.

#### Not sampled test:
In this scenario, we use the "correct value" of the TraceID, but mark it as not sampled, by adding "Sampled=0".
The expected outcome of this scenario is that the traces are not registered in X-Ray, but the TraceID is properly propagated upstream.

Request:
```
curl -v -H "x-amzn-trace-id: Root=1-62833b4c-4a0cb5dd511bf2cd48f99580;Sampled=0" vg.prod.local:8080/my-service/v1/test-endpoint
```

Got in "my-service": 
```
Root=1-62833e06-6c9b1bfca6444e9b803ec5e6;Parent=41201b3d10000f77;Sampled=0
```
Observation:
When `Sampled=0`, Envoys regenerate `X-Amzn-Trace-Id`, but keep the `Sampled=0`. This is completely unexpected, and breaks our log correlation.

### Route 2. Directly to "my-service" service.

The query:
```
curl -v -H "x-amzn-trace-id: <header value, see below>" my-service.dev.local/v1/test-endpoint
```

In this case, the nodes involved looked like this:
```
Client (Bastion) => Envoy (service sidecar) => service
```

Ran following requests:
```
curl -v -H "x-amzn-trace-id: Root=1-62833b4c-4a0cb5dd511bf2cd48f99580;Sampled=1" my-service.dev.local/v1/test-endpoint
curl -v -H "x-amzn-trace-id: Root=1-62833b4c-4a0cb5dd511bf2cd48f99580;Sampled=0" my-service.dev.local/v1/test-endpoint
```
In both cases the outcome was the same as with Virtual Gateway:
- When `Sampled=1`, Envoys used the provided TraceID, and the TraceID was propagated to the service.
- When `Sampled=0`, Envoys regenerated the TraceID, and passed a new value to the service, breaking the observability flow.

**Are you currently working around this issue?**
The only workaround we found is to reconfigure X-Ray / API GW Sampling rules to catch 100% requests, which is an expensive move and we'd rather avoid it.

**Additional context**
I have additionally reported that as an AWS Support Ticket with a bit more context and real service names. I'll link to this ticket in the AWS Console for correlation. I shall be able to disclose more information via AWS Support Ticket.

**Comment by suniltheta** [comment]
Hi @mkielar, thanks for raising this issue. Can you confirm if this issue is same as `Non Sampled X-Ray Tracing` [20127](https://github.com/envoyproxy/envoy/issues/20127) reported on envoyproxy github? The fix was already applied at `xray: non sampled trace should reuse trace id if set in header` [20400](https://github.com/envoyproxy/envoy/pull/20400).

The recently released AppMesh envoyproxy image v1.22 should contain the fix for this problem https://github.com/aws/aws-app-mesh-roadmap/issues/407. Let us know if using v1.22.0 image would fix your problem.

If you are not ready to upgrade to v1.22.0 at this time then we can consider including the fix as part of a patch release with v1.21.2

**Comment by mkielar** [comment]
Hi @suniltheta, thanks for the express response. 
Yes, this looks like the same as https://github.com/envoyproxy/envoy/issues/20127. I'll try upgrading our envoys to 1.22 and let you know.

**Comment by mkielar** [comment]
@suniltheta, one thing I found tricky about the migration is that you changed how the envoy docker images are built - previously they were based on AmazonLinux2, now they're "FROM scratch" images, meaning there's no "yum", etc. inside, and we need it to install some additional tools for health checks and boot-time initialization. 

I found this comment quite helpful: https://github.com/aws/aws-app-mesh-roadmap/issues/407#issuecomment-1122661513 but this kind of information should be made more easily accessible. May I suggest adding this piece of information here: https://docs.aws.amazon.com/app-mesh/latest/userguide/envoy.html?

**Comment by suniltheta** [comment]
Hi @mkielar we have released envoy v1.21.2 #412 with a fix for this bug. Let us know if your issue got resolved after using either v1.21.2 or v1.22.0

Meanwhile, we have updated our App Mesh docs at https://docs.aws.amazon.com/app-mesh/latest/userguide/envoy.html to include the information to generate a custom docker image on AL2, as per your above suggestion.

**Comment by mkielar** [comment]
Hi @suniltheta, 
We have are in a process of upgrading all our services to envoy `1.22.0` and the issue seems to be resolved for the ones running upgraded envoy. I can also see the documentation page has been updated with instructions on how to build Amazon Linux 2 based images, which is great!

Thanks a lot for express help. I think we can close the ticket now.
